File: tape/sdk/go/embedded/connectors.go
package embedded

// connectors.go — the three-method protocol every connector implements,
// plus the three result shapes. Mirrors `tape_adk/connectors.py`. A
// connector wraps ONE upstream (`bank.wire`, `payment.charge`) and is
// the only place in the system allowed to call it — the agent's tool
// body never does.

import (
	"context"
	"fmt"
	"sync"
)

// DispatchResult — what a single dispatch attempt produced. `Status` is
// "confirmed" | "unknown" | "failed". `RetryAfterMs` is a hint honored
// only when status="failed" (0 → use the dispatcher's default
// exponential backoff; negative → terminal FAILED).
type DispatchResult struct {
	Status       string
	ExternalRef  string
	Response     any
	Error        any
	RetryAfterMs int64
}

// ObservationResult — what `observe(business_key)` found upstream.
// `Status` is "confirmed" | "failed" | "absent" | "duplicate".
// `CompensateKind` is the obligation kind the reconciler should register
// atomically when status="duplicate".
type ObservationResult struct {
	Status         string
	ExternalRef    string
	Response       any
	Error          any
	CompensateKind string
}

// CompensationResult — what running the inverse produced. `Status` is
// "compensated" | "failed".
type CompensationResult struct {
	Status       string
	Response     any
	Error        any
	RetryAfterMs int64
}

// Connector — implement three methods and you can ride the reactor
// loops. `Name` is the registry key matching `effect.connector` (the
// same string an outbox tool declares at construction).
type Connector interface {
	Name() string
	Dispatch(ctx context.Context, effect EffectRecord) (DispatchResult, error)
	Observe(ctx context.Context, effect EffectRecord) (ObservationResult, error)
	Compensate(ctx context.Context, obligation ObligationRecord) (CompensationResult, error)
}

// LogConnector — a no-op connector that records every call. Useful for
// tests and demos. Always confirms on dispatch, reports absent on
// observe, and compensates successfully.
type LogConnector struct {
	NameStr string

	mu            sync.Mutex
	Dispatches    []EffectRecord
	Observations  []EffectRecord
	Compensations []ObligationRecord
}

// NewLogConnector — build a LogConnector with the given registry name
// (defaults to "log").
func NewLogConnector(name string) *LogConnector {
	if name == "" {
		name = "log"
	}
	return &LogConnector{NameStr: name}
}

func (l *LogConnector) Name() string { return l.NameStr }

func (l *LogConnector) Dispatch(_ context.Context, e EffectRecord) (DispatchResult, error) {
	l.mu.Lock()
	l.Dispatches = append(l.Dispatches, e)
	l.mu.Unlock()
	short := e.IdempotencyKey
	if len(short) > 8 {
		short = short[:8]
	}
	return DispatchResult{Status: "confirmed", ExternalRef: fmt.Sprintf("log-%s", short)}, nil
}

func (l *LogConnector) Observe(_ context.Context, e EffectRecord) (ObservationResult, error) {
	l.mu.Lock()
	l.Observations = append(l.Observations, e)
	l.mu.Unlock()
	return ObservationResult{Status: "absent"}, nil
}

func (l *LogConnector) Compensate(_ context.Context, o ObligationRecord) (CompensationResult, error) {
	l.mu.Lock()
	l.Compensations = append(l.Compensations, o)
	l.mu.Unlock()
	return CompensationResult{Status: "compensated"}, nil
}

File: tape/sdk/typescript/src/embedded/connectors.ts
// Connector protocol — how the outbox dispatcher talks to a counterparty,
// and how the reconciler asks the counterparty what really happened.
// Mirrors `tape_adk/connectors.py` shape-for-shape.
//
// A connector implements three async operations against one upstream system
// (`bank.wire`, `payment.charge`, `email.send`, …):
//
// * `dispatch(effect)` — actually call the upstream. Returns CONFIRMED,
//   UNKNOWN (call may have landed but the ack was lost), FAILED, or
//   throws (treated as retry-after-backoff).
// * `observe(effect)` — ask the upstream by `business_key` whether the
//   operation lives in its records. Returns CONFIRMED + external_ref,
//   FAILED, ABSENT, or DUPLICATE.
// * `compensate(obligation)` — run the inverse. Returns COMPENSATED or FAILED.

import type { EffectRecord, ObligationRecord } from './service.ts';

// ── result shapes ──────────────────────────────────────────────────────────

/** What a single dispatch attempt produced. */
export interface DispatchResult {
  /** 'confirmed' | 'unknown' | 'failed' */
  status: 'confirmed' | 'unknown' | 'failed';
  externalRef?: string;
  response?: unknown;
  error?: unknown;
  /** Backoff hint; only honored when status === 'failed'. 0 = default. */
  retryAfterMs?: number;
}

/** What observe(business_key) found on the counterparty's side. */
export interface ObservationResult {
  /** 'confirmed' | 'failed' | 'absent' | 'duplicate' | 'stuck' */
  status: 'confirmed' | 'failed' | 'absent' | 'duplicate' | 'stuck';
  externalRef?: string;
  response?: unknown;
  error?: unknown;
  /** When status === 'duplicate', the obligation kind to register. */
  compensateKind?: string;
}

/** What the inverse-operation call did. */
export interface CompensationResult {
  /** 'compensated' | 'failed' */
  status: 'compensated' | 'failed';
  response?: unknown;
  error?: unknown;
  retryAfterMs?: number;
}

// ── the protocol ───────────────────────────────────────────────────────────

/**
 * Implement three async methods and you're a connector. The reactor
 * library does the rest — claiming, transitioning, retrying with
 * backoff, and recording the result against the journal.
 *
 * `name` is the registry key — the same string used in `outboxTool({connector: 'bank.wire'})`.
 */
export interface Connector {
  name: string;
  dispatch(effect: EffectRecord): Promise<DispatchResult>;
  observe(effect: EffectRecord): Promise<ObservationResult>;
  compensate(obligation: ObligationRecord): Promise<CompensationResult>;
}

// ── tiny built-in connectors for tests + smoke runs ────────────────────────

/** A no-op connector that logs every call. Useful for tests + demos. */
export class LogConnector implements Connector {
  name: string;
  dispatches: EffectRecord[] = [];
  observations: EffectRecord[] = [];
  compensations: ObligationRecord[] = [];

  constructor(name = 'log') {
    this.name = name;
  }

  async dispatch(effect: EffectRecord): Promise<DispatchResult> {
    this.dispatches.push(effect);
    return {
      status: 'confirmed',
      externalRef: `log-${effect.idempotencyKey.slice(0, 8)}`,
    };
  }

  async observe(effect: EffectRecord): Promise<ObservationResult> {
    this.observations.push(effect);
    return { status: 'absent' };
  }

  async compensate(obligation: ObligationRecord): Promise<CompensationResult> {
    this.compensations.push(obligation);
    return { status: 'compensated' };
  }
}
